### PR TITLE
Pin surrogate-pair survival across read-block boundaries at every block size

### LINKED ISSUE #176: Add a test that divides a surrogate pair at a read-block boundary in the sanitizing reader
## 1. Summary

The class `XmlSanitizingTextReader` keeps surrogate pairs safe when a read boundary divides them. No test causes this condition. Add a test that causes this condition through the public API. This issue gives the full context for that task.

## 2. Background

PR #93 (commit `724883f7`) corrected a defect. Before that commit, the sanitizer removed surrogate pairs (emoji and other astral characters) from feed content.

The HTTP-to-XML rewrite then moved the load path to a new component. The load path does not call `RemoveInvalidXmlHexadecimalCharacters` now. The load path uses `XmlSanitizingTextReader` (`Solutions/Argotic.Common/XmlSanitizingTextReader.cs`). All the `CreateSafeNavigator` overloads in `SyndicationEncodingUtility` use this reader.

The reader removes the characters that XML does not permit. The reader keeps valid surrogate pairs. The reader has two hold slots (`XmlSanitizingTextReader.cs:46-53`):

- The slot `accepted` holds a character that the reader classified and approved.
- The slot `acceptedNext` holds only the low half of an approved pair.

When the caller's buffer becomes full after the high half, the low half stays in `acceptedNext`. The next read call sends the held low half to the caller with no new classification (`Dequeue`, `XmlSanitizingTextReader.cs:201-207`). This design prevents a specific defect. A low half that goes through classification again has no high half before it. The reader would then remove it as an unpaired surrogate. The output would contain a lone high surrogate, and the XML parser would stop with an error.

## 3. Problem

It is possible that no test causes a read boundary to fall between the two halves of a pair:

- The only consumer of the reader is `XmlReader`. `XmlReader` reads in large blocks.
- The dirty documents in `ParsePipelineDifferentialTests.TheTwoPipelines_AgreeOnEveryDirtyShape` are small. Each document fits in one block.
- The sample corpus is 7-bit ASCII (see the remark on `EverySampleDocument_ParsesIdenticallyThroughBothPipelines`). It contains no surrogate pairs.

Thus it is possible that `acceptedNext` never holds a value across two read calls in the full suite. The path is then code that no test examines.

**First step: measure.** Do not trust this analysis. Get branch coverage for `XmlSanitizingTextReader` and read the hit counts for `acceptedNext` and `Dequeue`. Use the procedure in `.endjin/build-warnings.md` §6.2. Obey the repository rule: measure before you change behaviour (CLAUDE.md, "Working Practice").

## 4. Constraints

- `XmlSanitizingTextReader` is internal. The test project has no `InternalsVisibleTo` grant. This is intentional. Write the test against the public API only.
- The public entry points are `SyndicationEncodingUtility.CreateSafeNavigator(string)`, `(Stream)`, `(Stream, Encoding)`, and `(TextReader)` (`SyndicationEncodingUtility.cs:327,381,442,521`). The resource types (`RssFeed.Load` and others) also apply.
- The block size of `XmlReader` is an implementation detail. Do not put the block size in the test. Do not quote a number that has no reproducible source in the repository (CLAUDE.md rule).
- Use Shouldly for the assertions.
- Put the test in `Solutions/Argotic.Extensions.Tests/Functionality/Common/`. `SanitiserCharacterisationTests.cs` is a possible location. A new file is also correct.
- Remove the trailing newline from each new compiled file (`Solutions/.editorconfig:9`).

## 5. Possible test design

A long run of astral characters removes the dependency on the block size:

1. Make a document that contains a text run of many surrogate pairs. For example, use 20,000 copies of U+1F600. This run is longer than one read block.
2. Make two documents: `<r>😀…😀</r>` and `<r>x😀…😀</r>`. The second document moves the run by one char position. For each possible block length, one of the two documents puts a block boundary between the two halves of some pair.
3. Parse each document through `CreateSafeNavigator(string)`, `CreateSafeNavigator(Stream)`, and `CreateSafeNavigator(TextReader)`.
4. Make sure the text value of the navigator is equal to the input run. Use full string equality. Equality finds a removed half, an added replacement char, and a changed order. A length check does not.

## 6. Make sure the test can fail

A verification that cannot fail is not a verification (CLAUDE.md, "Working Practice"; see also `.endjin/build-warnings.md` §13.5). Do these steps one time, before you trust the test:

1. Change `Dequeue` in a local worktree: remove the verbatim emission of `acceptedNext`, and send the low half through classification again.
2. Run the new test. Make sure the result is red.
3. Remove the local change. Run the test again. Make sure the result is green.

An alternative exists: add a committed negative control. `LegacyParsePipeline.Mutated` shows this pattern. The remark on `TheDifferential_ReportsADifference_WhenTheLegacyScanIsPerturbed` explains why a committed control is stronger than a one-time manual check.

## 7. Verification record

A session on 2026-08-10 confirmed that the PR #93 behaviour has no regression today:

- The 123 tests in `SyndicationEncodingUtilityTest` and `SanitiserCharacterisationTests` passed. The set includes the surrogate-pair cases.
- The 23 tests in `ParsePipelineDifferentialTests` passed. The negative control reported a difference, as required.
- An end-to-end script loaded an RSS document with emoji through `RssFeed.Load(Stream)`, saved it, and loaded it again. The emoji stayed. A U+0001 character was removed. Thus the sanitizer was active in the test.

That session found this gap. The gap is the only open item.

## 8. References

- PR #93: https://github.com/argotic-syndication-framework/Argotic/pull/93 (commit `724883f708ae5241532e5b4745ad7325f5f2e7d9`)
- `Solutions/Argotic.Common/XmlSanitizingTextReader.cs` — the two-slot design remark (lines 22-32), `Fill` (139-187), `Dequeue` (201-207)
- `Solutions/Argotic.Common/SyndicationEncodingUtility.cs:327-337` — the string overload wraps the sanitizing reader
- `Solutions/Argotic.Extensions.Tests/Functionality/Common/ParsePipelineDifferentialTests.cs` — the differential harness and its negative control
- `Solutions/Argotic.Extensions.Tests/TestDoubles/LegacyParsePipeline.cs` — the frozen legacy pipeline
- CLAUDE.md, "Working Practice" and "Quality Gates" — the rules named above

### PR DESCRIPTION
**Pull request by HowardvanRooijen**
Issue #176 asked for a test that divides a surrogate pair at a read-block
boundary in XmlSanitizingTextReader, and for the gap analysis to be measured
before being believed. Measured, the analysis is refuted: branch coverage on
the reader is saturated (Fill 16/16, Read(Span) 4/4), and a mutation that
reclassifies a low half held across a Read-call boundary — the exact defect
the two-slot design prevents — turns four existing tests red, all of them the
astral arm of ADirtyCharacterAtEveryBufferBoundary_SanitisesIdentically,
failing at position 4097.

So the condition is exercised today, but only because the sweep's offset
brackets straddle today's 4,096-char block. The new row removes that
dependency: a run of 20,000 astral characters is longer than any plausible
read block, and of two runs starting one position apart, one must place a
block boundary between the halves of some pair whatever the block length is.
Asserted with full string equality through all four public entry points.

Verified per the issue's protocol: green against unmodified code, red under
the boundary mutation (1/1 failed), green again after revert; whole offline
suite 3,551/3,551.

Closes #176

Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: Solutions/Argotic.Extensions.Tests/Functionality/Common/SanitiserCharacterisationTests.cs [hunk 1]
```diff
@@ -269,6 +269,49 @@ public void ADirtyCharacterAtEveryBufferBoundary_SanitisesIdentically(string ent
...
+        foreach (string prefix in alignments)
+        {
+            string document = string.Concat("<r>", prefix, run, "</r>");
+
+            foreach (string entryPoint in EntryPoints)
+            {
+                RootOf(Parse(entryPoint, document))
+                    .ShouldBe(document, $"{entryPoint}, run at offset {prefix.Length}");
+            }
+        }
```
[the remark this answers is not in the document]
Comment (HowardvanRooijen) [reply Solutions/Argotic.Extensions.Tests/Functionality/Common/SanitiserCharacterisationTests.cs]: Good catch — the shared `EntryPoints` list covers four call shapes but not `CreateSafeNavigator(Stream, Encoding)`. Fixed in f6c15b75: the test now iterates the shared list plus a `stream+encoding` case wired through the same `Parse` helper.

Verified the addition can fail, not just that it passes: under the held-low-half mutation used for the original red/green protocol, the test is red through the `stream+encoding` shape *alone*, so the overload genuinely traverses the sanitising reader rather than passing incidentally. Clean, the offline suite is 3,551/3,551.

**Review by JamesDawson — APPROVED** [review]
